sql/core/src/main/scala/org/apache/spark/sql/sources/commands.scala

```diff
@@ -218,24 +218,20 @@ private[sql] case class InsertIntoHadoopFsRelation(
         val partitionProj = newProjection(codegenEnabled, partitionOutput, output)
         val dataProj = newProjection(codegenEnabled, dataOutput, output)
 
-        if (needsConversion) {
-          val converter = CatalystTypeConverters.createToScalaConverter(dataSchema)
-          while (iterator.hasNext) {
-            val row = iterator.next()
-            val partitionPart = partitionProj(row)
-            val dataPart = dataProj(row)
-            val convertedDataPart = converter(dataPart).asInstanceOf[Row]
-            writerContainer.outputWriterForRow(partitionPart).write(convertedDataPart)
-          }
+        val partitionSchema = StructType.fromAttributes(partitionOutput)
+        val partConverter = CatalystTypeConverters.createToScalaConverter(partitionSchema)
```

Comment:
In the old code, it looks like we'd only create a converter from the `partitionSchema` in cases where `needsConversion` was `false`, but here it looks like we do it in both cases.  Can you explain why you made this change?